### PR TITLE
fix: add response_type=token+id_token for ROPC flow

### LINKED ISSUE #3883: acquireTokenByUsernamePassword (ROPC) will not return an id_token in Azure AD B2C
### Core Library

MSAL Node (@azure/msal-node)

### Core Library Version

1.2.0

### Wrapper Library

Not Applicable

### Wrapper Library Version

None

### Description

When using the Resource Owner Password Credential Flow with Azure AD B2C an id_token seems to be only returned when:

- scopes include `openid`
- response_type is `id_token` or `token id_token`

See [Test the ROPC Flow](https://docs.microsoft.com/en-us/azure/active-directory-b2c/add-ropc-policy?tabs=app-reg-ga&pivots=b2c-user-flow#test-the-ropc-flow) in the Azure AD B2C Doumentation for a sample request returning both `access_token` and `id_token`:

**Request:**

```
POST /<tenant-name>.onmicrosoft.com/B2C_1A_ROPC_Auth/oauth2/v2.0/token HTTP/1.1
Host: <tenant-name>.b2clogin.com
Content-Type: application/x-www-form-urlencoded

username=contosouser.outlook.com.ws&password=Passxword1&grant_type=password&scope=openid+bef22d56-552f-4a5b-b90a-1988a7d634ce+offline_access&client_id=bef22d56-552f-4a5b-b90a-1988a7d634ce&response_type=token+id_token
```

**Response**

```
{
    "access_token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsImtpZCI6Ik9YQjNhdTNScWhUQWN6R0RWZDM5djNpTmlyTWhqN2wxMjIySnh6TmgwRlki...",
    "token_type": "Bearer",
    "expires_in": "3600",
    "refresh_token": "eyJraWQiOiJacW9pQlp2TW5pYVc2MUY0TnlfR3REVk1EVFBLbUJLb0FUcWQ1ZWFja1hBIiwidmVyIjoiMS4wIiwiemlwIjoiRGVmbGF0ZSIsInNlciI6Ij...",
    "id_token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsImtpZCI6Ik9YQjNhdTNScWhUQWN6R0RWZDM5djNpTmlyTWhqN2wxMjIySnh6TmgwRlki..."
}
````
When using acquireTokenByUsernamePassword the `response_type` parameter does not appear to be set. Therefore, no `id_token` will be returned by Azure AD B2C, even when list of scopes include `openid`

Expected behaviour:
- set the `response_type` in acquireTokenByUsernamePassword to `token id_token` when the scopes include `openind`

### Error Message

_No response_

### Msal Logs

_No response_

### MSAL Configuration

```javascript
auth: {
        clientId: "clientID",
        authority: "https://yourtenant.b2clogin.com/yourtenant.onmicrosoft.com/b2c_1a_signinwithropc", 
        knownAuthorities: ["https://yourtenant.b2clogin.com"]
      },
      cache: {
          cacheLocation: "sessionStorage",
          storeAuthStateInCookie: false
      }
```

### Relevant Code Snippets

```javascript
import * as msal from "@azure/msal-node"

const username = "email@example.com"
const password = "superSecret"

   const config = {
     auth: {
        clientId: "clientID",
        authority: "https://yourtenant.b2clogin.com/yourtenant.onmicrosoft.com/b2c_1a_signinwithropc", 
        knownAuthorities: ["https://yourtenant.b2clogin.com"]
      },
      cache: {
          cacheLocation: "sessionStorage",
          storeAuthStateInCookie: false
      }

    }
    
const msalInstance = new msal.PublicClientApplication(config);
         
const usernamePasswordRequest = {
            scopes: ["clientId", "openid"],
            username: username,
            password: password
        };
      
msalInstance.acquireTokenByUsernamePassword(usernamePasswordRequest).then(response => {
            console.log("acquired token by password grant");
        }).catch((error) => {
          console.log(error);
      });
```

### Reproduction Steps

1. request both access_token and id_token by calling acquireTokenByUsernamePassword with the following scopes: 

```
scopes: ["clientId", "openid"]
```

### Expected Behavior

- acquireTokenByUsernamePassword should set the response_type to `token id_token` when the scopes include `openind`
- get both `access_token` and `id_token` in the response

### Identity Provider

Azure B2C Basic Policy

### Browsers Affected (Select all that apply)

Chrome

### Regression

_No response_

### Source

External (Customer)

### PR DESCRIPTION
**Pull request by ckumpe**
This should fix https://github.com/AzureAD/microsoft-authentication-library-for-js/issues/3883

### CODE REVIEW

File: lib/msal-common/src/request/RequestParameterBuilder.ts [hunk 1]
```diff
@@ -30,6 +30,15 @@ export class RequestParameterBuilder {
         );
     }
 
+    /**
+     * add response_type = code
+     */
+    addResponseTypeForIdToken(): void {
+        this.parameters.set(
+            AADServerParamKeys.RESPONSE_TYPE, encodeURIComponent(Constants.TOKEN_RESPONSE_TYPE + " " + Constants.ID_TOKEN_RESPONSE_TYPE)
```
Comment (jasonnutter) [inline lib/msal-common/src/request/RequestParameterBuilder.ts]:
```suggestion
            AADServerParamKeys.RESPONSE_TYPE, encodeURIComponent(`${Constants.TOKEN_RESPONSE_TYPE} ${Constants.ID_TOKEN_RESPONSE_TYPE}`);
```

File: lib/msal-common/src/request/RequestParameterBuilder.ts [hunk 2]
```diff
@@ -30,6 +30,15 @@ export class RequestParameterBuilder {
         );
     }
 
+    /**
+     * add response_type = code
+     */
+    addResponseTypeForIdToken(): void {
```
Comment (jasonnutter) [inline lib/msal-common/src/request/RequestParameterBuilder.ts]:
```suggestion
    addResponseTypeForTokenIdToken(): void {
```

**Comment by jasonnutter** [comment]
@ckumpe Thanks for opening this! Can you please show a screenshot/logs/etc for doing ROPC for B2C before and after this change to confirm it is working? Furthermore, please run `npm run beachball:change` from the root of the repo and push up the generated change file, thanks!

File: lib/msal-common/src/request/RequestParameterBuilder.ts [hunk 3]
```diff
@@ -30,6 +30,15 @@ export class RequestParameterBuilder {
         );
     }
 
+    /**
+     * add response_type = code
```
Comment (jasonnutter) [inline lib/msal-common/src/request/RequestParameterBuilder.ts]:
```suggestion
     * add response_type = token id_token
```

**Comment by ckumpe** [comment]
I executed the requested tests to confirm the fix is working.

### Test setup
I've executed this simple test which retrieves a token with `acquireTokenByUsernamePassword` and afterwards try to use token cache and retrieve a token by `acquireTokenSilent`.

```typescript
import { PublicClientApplication, SilentFlowRequest, UsernamePasswordRequest } from '@azure/msal-node';
import { Configuration } from '@azure/msal-node/dist/config/Configuration';

const username = "...";
const password = "...";
const tenant = "...";
const clientId = "...";

const config = {
  auth: {
    clientId: clientId,
    authority: `https://${tenant}.b2clogin.com/${tenant}.onmicrosoft.com/b2c_1_signinropc`,
    knownAuthorities: [`https://${tenant}.b2clogin.com`]
  }
} as Configuration;

const msalInstance = new PublicClientApplication(config);

const usernamePasswordRequest = {
  scopes: [clientId, "openid"],
  username: username,
  password: password
} as UsernamePasswordRequest;

const response = await msalInstance.acquireTokenByUsernamePassword(usernamePasswordRequest);
console.log("acquired token by password grant:", response);

// These checks will fail wihout the bugfix
if (!response?.idToken) throw Error("response does not contain an ID token");
if (!response?.account) throw Error("response does not contain an account required for acquireTokenSilent");

const silentRequest = {
  account: response.account,
  scopes: [clientId, "openid"],
} as SilentFlowRequest;

const authenticationResult =  await msalInstance.acquireTokenSilent(silentRequest);
console.log("acquired token by silent flow:", response);
```

### Before
This test will fail because of the missing ID token in the response of `acquireTokenByUsernamePassword`:
```
acquired token by password grant: {
  authority: 'https://yourtenant.b2clogin.com/yourtenant.onmicrosoft.com/b2c_1_signinropc/',
  scopes: [
    '...',
    'openid',
    'profile',
    'offline_access'
  ],
  account: null,
  idToken: '',
  accessToken: 'eyJ0eXAiOiJKV1Qi...',
  expiresOn: 2022-04-12T09:12:50.000Z,
  correlationId: 'c1418207-ae1f-48ed-9bc8-4835118363c7',
  ...
}
Error: response does not contain an ID token
    at .../auth2.ts:31:31
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
```

### After
After the bugfix `acquireTokenByUsernamePassword` returns an ID token and the test will continue and can successfully use `acquireTokenSilent`:
```
acquired token by password grant: {
  authority: 'https://yourtenant.b2clogin.com/yourtenant.onmicrosoft.com/b2c_1_signinropc/',
  scopes: [
    '...',
    'openid',
    'profile',
    'offline_access'
  ],
  account: {
    homeAccountId: '...',
    environment: 'yourtenant.b2clogin.com',
    username: '...',
    localAccountId: '...',
    ...
  },
  idToken: 'eyJ0eXAiOiJKV1Qi...',
  accessToken: 'eyJ0eXAiOi...',
  expiresOn: 2022-04-12T09:07:31.000Z,
  correlationId: '659eb3fd-7cdb-4306-b079-2c6786376403',
  ...
}
acquired token by silent flow: {
  authority: 'https://yourtenant.b2clogin.com/yourtenant.onmicrosoft.com/b2c_1_signinropc/',
  scopes: [
    '...',
    'openid',
    'profile',
    'offline_access'
  ],
  account: {
    homeAccountId: '...',
    environment: 'yourtenant.b2clogin.com',
    username: '...',
    localAccountId: '...',
    ...
  },
  idToken: 'eyJ0eXAiOi...',
  accessToken: 'eyJ0eXAiOi...',
  expiresOn: 2022-04-12T09:07:31.000Z,
  correlationId: '659eb3fd-7cdb-4306-b079-2c6786376403',
  ...
}
```

**Comment by jasonnutter** [comment]
@ckumpe Perfect, thanks!

File: change/@azure-msal-common-d22257a2-f552-4172-b1be-681ce2fa0768.json [hunk 4]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "patch",
+  "comment": "acquireTokenByUsernamePassword will now return an id_token in Azure AD B2C (fixes #3883)",
```
Comment (jasonnutter) [inline change/@azure-msal-common-d22257a2-f552-4172-b1be-681ce2fa0768.json]:
```suggestion
  "comment": "acquireTokenByUsernamePassword will now return an id_token in Azure AD B2C #4694",
```

**Review by jasonnutter — APPROVED** [review]

**Review by tnorling — APPROVED** [review]

**Pull request merged**